### ISSUE TITLE
Baseline Implementations

### ISSUE THREAD

**Issue by bakszero** [microsoft/MASS#11]
Hi, thanks for sharing this work! Would really appreciate if you could point me out to the baseline implementations for unsupervised NMT -- specifically the BERT+LM and DAE methods with perhaps more details on the experimental parameters for the baseline tasks --  since I couldn't find them in the paper. Thanks.

**Comment by StillKeepTry** [comment]
For bert (mlm) + clm, [lample's method](https://arxiv.org/abs/1901.07291) has implemented. The fine-tune command is as:
```
wget -c https://dl.fbaipublicfiles.com/XLM/mlm_enfr_1024.pth
wget -c https://dl.fbaipublicfiles.com/XLM/clm_enfr_1024.pth

encoder=mlm_enfr_1024.pth
decoder=mlm_enfr_1024.pth

python train.py \
--exp_name unsupMT_enfr   \
--reload_model '$encoder,$decoder'  \
--data_path ./data/processed/en-fr/  \
--lgs 'en-fr'                            \
--ae_steps 'en,fr'                   \
--bt_steps 'en-fr-en,fr-en-fr' \
--word_shuffle 3                   \
--word_dropout 0.1              \
--word_blank 0.1                  \
--lambda_ae '0:1,100000:0.1,300000:0'  \
--encoder_only false            \
--emb_dim 1024                  \
--n_layers 6                          \
--n_heads 8                          \
--dropout 0.1                       \
--attention_dropout 0.1       \
--gelu_activation true          \
--tokens_per_batch 2000     \
--batch_size 32                    \
--bptt 256                            \
--optimizer adam_inverse_sqrt,beta1=0.9,beta2=0.98,lr=0.0001 \
--epoch_size 200000           \
--eval_bleu true                   \
```

For DAE, you can train a DAE pre-trained model by using:
```
python train.py \
--exp_name unsupMT_enfr   \
--data_path ./data/processed/en-fr/  \
--lgs 'en-fr'                            \
--ae_steps 'en,fr'                   \
--bt_steps 'en-fr-en,fr-en-fr' \
--lambda_bt '0:0,10000:0'    \
--word_shuffle 3                   \
--word_dropout 0.1              \
--word_blank 0.1                  \
--lambda_ae '0:1,100000:0.1,300000:0'  \
--encoder_only false            \
--emb_dim 1024                  \
--n_layers 6                          \
--n_heads 8                          \
--dropout 0.1                       \
--attention_dropout 0.1       \
--gelu_activation true          \
--tokens_per_batch 2000     \
--batch_size 32                    \
--bptt 256                            \
--optimizer adam_inverse_sqrt,beta1=0.9,beta2=0.98,lr=0.0001 \
--epoch_size 200000           \
--eval_bleu true                   \
```
As DAE may produce worse results, i suggest you can load a mlm/clm model or a pre-trained word embedding for model initial when using DAE pre-training. (DAE will easily cause copy mechanism).

**Comment by bakszero** [comment]
Thankyou very much! Loading the pretrained encoder gives OOM issue on a 16 GB GPU results in process killed. Do you believe there might be a way to circumvent this?

**Comment by StillKeepTry** [comment]
@bakszero how many tokens do you use? if `tokens_per_batch` is 2000, I think it can be handled in a 16GB V100. If it is still out-of-memory, you can reduce `tokens_per_batch` to have a try.

**Comment by weiyengs** [comment]
@StillKeepTry 
Thanks for the hints! 

Does the same codes apply to the text summarization experiment in the paper too?

Thanks!

**Comment by bakszero** [comment]
> @bakszero how many tokens do you use? if `tokens_per_batch` is 2000, I think it can be handled in a 16GB V100. If it is still out-of-memory, you can reduce `tokens_per_batch` to have a try.

Thanks @StillKeepTry for your suggestion. However, even when lowering `tokens_per_batch` to the lowest setting, I'm unable to load it onto the GPU due to memory constraints. After multiple experimentations with a lot of different parameters, I realised reducing the embedding size to 512 is what does the trick -- for training from scratch as well as ignoring GPU errors. Unfortunately, there is no 512-dim trained model available anywhere. If you do have access to a 512-dim trained model, please let me know, it'll be of great help!
